File: crypto/BRCryptoHasher.c
//
//  BRCryptoHasher.c
//  BRCore
//
//  Created by Michael Carrara on 9/23/19.
//  Copyright © 2019 Breadwinner AG. All rights reserved.
//
//  See the LICENSE file at the project root for license information.
//  See the CONTRIBUTORS file at the project root for a list of contributors.

#include <assert.h>
#include <stdlib.h>

#include "BRCryptoHasher.h"
#include "support/BRCrypto.h"

struct BRCryptoHasherRecord {
    BRCryptoHasherType type;
    BRCryptoRef ref;
};

IMPLEMENT_CRYPTO_GIVE_TAKE (BRCryptoHasher, cryptoHasher);

extern BRCryptoHasher
cryptoHasherCreate(BRCryptoHasherType type) {
    BRCryptoHasher hasher = NULL;

    switch (type) {
        case CRYPTO_HASHER_SHA1:
        case CRYPTO_HASHER_SHA224:
        case CRYPTO_HASHER_SHA256:
        case CRYPTO_HASHER_SHA256_2:
        case CRYPTO_HASHER_SHA384:
        case CRYPTO_HASHER_SHA512:
        case CRYPTO_HASHER_SHA3:
        case CRYPTO_HASHER_RMD160:
        case CRYPTO_HASHER_HASH160:
        case CRYPTO_HASHER_KECCAK256:
        case CRYPTO_HASHER_MD5: {
            hasher = calloc (1, sizeof(struct BRCryptoHasherRecord));
            hasher->type = type;
            hasher->ref = CRYPTO_REF_ASSIGN(cryptoHasherRelease);
            break;
        }
        default: {
            // for an unsupported algorithm, assert
            assert (0);
            break;
        }
    }

    return hasher;
}

static void
cryptoHasherRelease (BRCryptoHasher hasher) {
    memset (hasher, 0, sizeof(*hasher));
    free (hasher);
}

extern size_t
cryptoHasherLength (BRCryptoHasher hasher) {
    size_t length = 0;

    switch (hasher->type) {
        case CRYPTO_HASHER_SHA1: {
            length = 20;
            break;
        }
        case CRYPTO_HASHER_SHA224: {
            length = 28;
            break;
        }
        case CRYPTO_HASHER_SHA256: {
            length = 32;
            break;
        }
        case CRYPTO_HASHER_SHA256_2: {
            length = 32;
            break;
        }
        case CRYPTO_HASHER_SHA384: {
            length = 48;
            break;
        }
        case CRYPTO_HASHER_SHA512: {
            length = 64;
            break;
        }
        case CRYPTO_HASHER_SHA3: {
            length = 32;
            break;
        }
        case CRYPTO_HASHER_RMD160: {
            length = 20;
            break;
        }
        case CRYPTO_HASHER_HASH160: {
            length = 20;
            break;
        }
        case CRYPTO_HASHER_KECCAK256: {
            length = 32;
            break;
        }
        case CRYPTO_HASHER_MD5: {
            length = 16;
            break;
        }
        default: {
            // for an unsupported algorithm, assert
            assert (0);
            break;
        }
    }

    return length;
}

extern BRCryptoBoolean
cryptoHasherHash (BRCryptoHasher hasher,
                  uint8_t *dst,
                  size_t dstLen,
                  const uint8_t *src,
                  size_t srcLen) {
    // - src CAN be NULL, if srcLen is 0
    // - dst MUST be non-NULL and sufficiently sized
    if ((NULL == src && 0 != srcLen) ||
        NULL == dst || dstLen < cryptoHasherLength (hasher)) {
        assert (0);
        return CRYPTO_FALSE;
    }

    // the hash routines don't return anything so assume success
    // and only treat the unhandled case as failure
    BRCryptoBoolean result = CRYPTO_TRUE;

    switch (hasher->type) {
        case CRYPTO_HASHER_SHA1: {
            BRSHA1 (dst, src, srcLen);
            break;
        }
        case CRYPTO_HASHER_SHA224: {
            BRSHA224 (dst, src, srcLen);
            break;
        }
        case CRYPTO_HASHER_SHA256: {
            BRSHA256 (dst, src, srcLen);
            break;
        }
        case CRYPTO_HASHER_SHA256_2: {
            BRSHA256_2 (dst, src, srcLen);
            break;
        }
        case CRYPTO_HASHER_SHA384: {
            BRSHA384 (dst, src, srcLen);
            break;
        }
        case CRYPTO_HASHER_SHA512: {
            BRSHA512 (dst, src, srcLen);
            break;
        }
        case CRYPTO_HASHER_SHA3: {
            BRSHA3_256 (dst, src, srcLen);
            break;
        }
        case CRYPTO_HASHER_RMD160: {
            BRRMD160 (dst, src, srcLen);
            break;
        }
        case CRYPTO_HASHER_HASH160: {
            BRHash160 (dst, src, srcLen);
            break;
        }
        case CRYPTO_HASHER_KECCAK256: {
            BRKeccak256 (dst, src, srcLen);
            break;
        }
        case CRYPTO_HASHER_MD5: {
            BRMD5 (dst, src, srcLen);
            break;
        }
        default: {
            // for an unsupported algorithm, assert
            assert (0);
            result = CRYPTO_FALSE;
            break;
        }
    }

    return result;
}

File: Java/CoreNative/src/main/java/com/breadwallet/corenative/crypto/BRCryptoHasher.java
/*
 * Created by Michael Carrara <michael.carrara@breadwallet.com> on 7/1/19.
 * Copyright (c) 2019 Breadwinner AG.  All right reserved.
*
 * See the LICENSE file at the project root for license information.
 * See the CONTRIBUTORS file at the project root for a list of contributors.
 */
package com.breadwallet.corenative.crypto;

import com.breadwallet.corenative.CryptoLibraryDirect;
import com.breadwallet.corenative.utility.SizeT;
import com.google.common.base.Optional;
import com.google.common.primitives.Ints;
import com.sun.jna.Pointer;
import com.sun.jna.PointerType;

public class BRCryptoHasher extends PointerType {

    // these must mirror BRCryptoHasherType's enum values
    private static final int CRYPTO_HASHER_SHA1         = 0;
    private static final int CRYPTO_HASHER_SHA224       = 1;
    private static final int CRYPTO_HASHER_SHA256       = 2;
    private static final int CRYPTO_HASHER_SHA256_2     = 3;
    private static final int CRYPTO_HASHER_SHA384       = 4;
    private static final int CRYPTO_HASHER_SHA512       = 5;
    private static final int CRYPTO_HASHER_SHA3         = 6;
    private static final int CRYPTO_HASHER_RMD160       = 7;
    private static final int CRYPTO_HASHER_HASH160      = 8;
    private static final int CRYPTO_HASHER_KECCAK256    = 9;
    private static final int CRYPTO_HASHER_MD5          = 10;

    public static Optional<BRCryptoHasher> createSha1() {
        return create(CRYPTO_HASHER_SHA1);
    }

    public static Optional<BRCryptoHasher> createSha224() {
        return create(CRYPTO_HASHER_SHA224);
    }

    public static Optional<BRCryptoHasher> createSha256() {
        return create(CRYPTO_HASHER_SHA256);
    }

    public static Optional<BRCryptoHasher> createSha256_2() {
        return create(CRYPTO_HASHER_SHA256_2);
    }

    public static Optional<BRCryptoHasher> createSha384() {
        return create(CRYPTO_HASHER_SHA384);
    }

    public static Optional<BRCryptoHasher> createSha512() {
        return create(CRYPTO_HASHER_SHA512);
    }

    public static Optional<BRCryptoHasher> createSha3() {
        return create(CRYPTO_HASHER_SHA3);
    }

    public static Optional<BRCryptoHasher> createRmd160() {
        return create(CRYPTO_HASHER_RMD160);
    }

    public static Optional<BRCryptoHasher> createHash160() {
        return create(CRYPTO_HASHER_HASH160);
    }

    public static Optional<BRCryptoHasher> createKeccak256() {
        return create(CRYPTO_HASHER_KECCAK256);
    }

    public static Optional<BRCryptoHasher> createMd5() {
        return create(CRYPTO_HASHER_MD5);
    }

    private static Optional<BRCryptoHasher> create(int alg) {
        return Optional.fromNullable(CryptoLibraryDirect.cryptoHasherCreate(alg)).transform(BRCryptoHasher::new);
    }

    public BRCryptoHasher() {
        super();
    }

    public BRCryptoHasher(Pointer address) {
        super(address);
    }

    public Optional<byte[]> hash(byte[] data) {
        Pointer thisPtr = this.getPointer();

        SizeT length = CryptoLibraryDirect.cryptoHasherLength(thisPtr);
        int lengthAsInt = Ints.checkedCast(length.longValue());
        if (0 == lengthAsInt) return Optional.absent();

        byte[] hash = new byte[lengthAsInt];
        int result = CryptoLibraryDirect.cryptoHasherHash(thisPtr, hash, new SizeT(hash.length), data, new SizeT(data.length));
        return result == BRCryptoBoolean.CRYPTO_TRUE ? Optional.of(hash) : Optional.absent();
    }

    public void give() {
        Pointer thisPtr = this.getPointer();

        CryptoLibraryDirect.cryptoHasherGive(thisPtr);
    }
}
